### DISCUSSION TITLE
Limit download of files and sequences

### DISCUSSION THREAD

**Discussion by andreaminio** [GMOD/jbrowse-components#3143] [General] [answered]
Hi everyone,

I'm trying to setup the new Jbrowse2 for our lab, however, since some of the browsers to our data are released in advance before publication, I have the necessity to limit/control the download of files and sequences from within the browser.
So I'd need to control: 

- The sequence download using the "Get sequence" option on the zoom in menu. Possibly limiting the download to only to selected regions up to a given length (And not potentially the whole chromosome) and, eventually, add an optional link to file download page.
- Fields in the info panel of each track. I'd like to remove from the features "_something_Location" that do link to the internal file(s) loaded in the track (ex. _adapter.gffGzLocation_, _adapter.pafLocation_), eventually substituting it with an external link (like for the feature panel) and a copyright/citation info.

I suppose this is possible through the configuration file on each track or genome, like for the feature panel content manipulation, but I have hard time finding how.

Thanks in advance,

Andrea

**Comment by garrettjstevens** [comment]
These things might be possible, but they would require writing a custom plugin. I can give some guidance for how to get started with that, but I want to make sure that you are aware of the limitations first. You can block mentions of the file names and downloads in JBrowse, but that does not secure your data at all. Someone could still easily check the file URL of what JBrowse is accessing in their browser's developer tools and then download the whole file. So what you are proposing would discourage downloads, but not prevent them.

**Reply by andreaminio** [reply]
Hi Garrett,

Thanks for the fast reply! 

I see the point...well if I can't avoid completely exposing the files discourage downloads seems, at least, better than nothing. At least not having the links in plain sight will help. 

Please help me hiding the plain sight links at least. For adding custom links I may find a way later.

May it be a possible to use htaccess to block the "direct downloads" of the files only while JBrowse may still be able to access the files within the server?

Andrea

**Reply by garrettjstevens** [reply]
I don't know of a good way to do that with htaccess. You could perhaps try only allowing range requests for the data files (you'd still need regular access for the index files, though), but that might still break something in JBrowse.

For the customization, you can start a plugin from the template here: https://github.com/GMOD/jbrowse-plugin-template. To override the "Get sequence", you could try something like the track menu items here: https://jbrowse.org/jb2/docs/developer_guide/#adding-track-context-menu-items. Instead of "LinearPileupDisplay" you'd use "LinearGenomeView", and instead of "contextMenuItems" you'd use "rubberBandMenuItems". See the current implementation of that menu item here: https://github.com/GMOD/jbrowse-components/blob/fcbd7611478db59e50bbd426ddc9e3fffaaac33b/plugins/linear-genome-view/src/LinearGenomeView/index.tsx#L1164-L1170

You might be able to do what you need with the feature details panel using track configuration instead of a plugin. See here for more info: https://jbrowse.org/jb2/docs/config_guide/#customizing-feature-details-panels.

**Comment by cmdcolin** [comment, marked as answer]
We have the hideUris setting for the "about dialog" now. Example config showing the usage of this setting:

```
{
      "type": "VariantTrack",
      "trackId": "volvox_test_vcf",
      "formatAbout": {
        "hideUris": true
      },
      "name": "volvox 1000genomes vcf",
      "assemblyNames": ["volvox"],
      "adapter": {
        "type": "VcfTabixAdapter",
        "vcfGzLocation": {
          "uri": "volvox.test.vcf.gz"
        },
        "index": {
          "location": {
            "uri": "volvox.test.vcf.gz.tbi"
          }
        }
      }
    },
```
